### PR TITLE
PIN-3007 [BE] Creazione di un job che recuperi tutte le privacy notice presenti su OneTrust ogni 5 minuti

### PR DESCRIPTION
**Pull request by nttdata-rtorsoli**

### CODE REVIEW

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/error/PrivacyNoticeError.scala [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+      extends ComponentError("0002", s"Error while reading data from One Trust with status code $statusCode")
+  final case class OneTrustAuthError(error: OneTrustAuthenticationError, statusCode: StatusCode)
+      extends ComponentError(
+        "0003",
+        s"Error while reading data from One Trust with message ${error.error}, description${error.error_description} status code $statusCode"
+      )
+  final case class OneTrustDeserializationError(body: String, error: Exception)
+      extends ComponentError(
+        "0004",
+        s"Error while reading data from One Trust -> $body and error ${error.getMessage()}"
```
Comment (galales) [inline privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/error/PrivacyNoticeError.scala]: after your tests, do you know if OneTrust may return sensitive information? If positive, maybe we could log just the error message, and not the body.

File: privacy-notices-updater/src/main/resources/application.conf [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+dynamo {
+    table-name = ${PRIVACY-NOTICES_DYNAMO_TABLE_NAME}
+}
+one-trust {
+    token-url           = ${PRIVACY-NOTICES_ONE_TRUST_TOKEN_URL}
+    privacy-notices-url = ${PRIVACY-NOTICES_ONE_TRUST_PN_URL}
+    pp-uuid             = ${PRIVACY-NOTICES_PP_UUID}
+    tos-uuid            = ${PRIVACY-NOTICES_TOS_UUID}
+    client-id           = ${PRIVACY-NOTICES_CLIENT_ID}
+    client-secret       = ${PRIVACY-NOTICES_CLIENT_SECRET}
```
Comment (galales) [inline privacy-notices-updater/src/main/resources/application.conf]: to keep naming convention, could you change `PRIVACY-NOTICES_*` to `PRIVACY_NOTICES_*`?

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/error/PrivacyNoticeError.scala [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+object PrivacyNoticeError {
+  final case class DynamoReadingError(message: String)
+      extends ComponentError("0001", s"Error while reading data from Dynamo -> $message")
+  final case class OneTrustHttpError(statusCode: StatusCode)
+      extends ComponentError("0002", s"Error while reading data from One Trust with status code $statusCode")
+  final case class OneTrustAuthError(error: OneTrustAuthenticationError, statusCode: StatusCode)
+      extends ComponentError(
+        "0003",
+        s"Error while reading data from One Trust with message ${error.error}, description${error.error_description} status code $statusCode"
```
Comment (galales) [inline privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/error/PrivacyNoticeError.scala]:
```suggestion
        s"Error while reading data from One Trust with message ${error.error}, description ${error.error_description} status code $statusCode"
```

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/DynamoService.scala [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  override def getById(id: UUID): Future[Option[PrivacyNotice]] = {
+    logger.info(s"Getting id $id privacy notice")
+    scanamo
+      .exec { table.get("id" === id) }
+      .flatMap {
+        case Some(value) =>
+          value match {
+            case Right(x)  => Future.successful(x.some)
+            case Left(err) => Future.failed(DynamoReadingError(describe(err)))
+          }
```
Comment (galales) [inline privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/DynamoService.scala]: FYI we have a `toFuture()` method in utils for this case, like
```suggestion
        case Some(value) => value.leftMap(err => DynamoReadingError(describe(err)).toFuture
```

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/OneTrustService.scala [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+final class OneTrustServiceImpl(config: OneTrustConfiguration)(implicit
+  ec: ExecutionContext,
+  actorSystem: ActorSystem,
+  contexts: Seq[(String, String)]
+) extends OneTrustService {
+
+  implicit val logger: LoggerTakingImplicit[ContextFieldsToLog] =
+    Logger.takingImplicit[ContextFieldsToLog](this.getClass)
+
+  val backend = Slf4jLoggingBackend(AkkaHttpBackend.usingActorSystem(actorSystem))
```
Comment (galales) [inline privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/OneTrustService.scala]: if this is the only use of akka, we could use a different backend and remove this dependency

**Review by galales — COMMENTED** [review]
the build is failing because a test in a job requires the new field `kind` in tenant. May I ask you to fix it?

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/error/PrivacyNoticeError.scala (hunk 1 above)
Comment (nttdata-rtorsoli) [reply privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/error/PrivacyNoticeError.scala]: ok

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/OneTrustService.scala (hunk 5 above)
Comment (nttdata-rtorsoli) [reply privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/OneTrustService.scala]: Which kind of backend do you propose?
https://sttp.softwaremill.com/en/latest/backends/summary.html

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/DynamoService.scala (hunk 4 above)
Comment (nttdata-rtorsoli) [reply privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/DynamoService.scala]: value.leftMap(err => DynamoReadingError(describe(err))).toFuture.some.sequence, it is good for you?

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/OneTrustService.scala (hunk 5 above)
Comment (galales) [reply privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/OneTrustService.scala]: Probably in this case a simple `DefaultFutureBackend` is enough?

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/DynamoService.scala (hunk 4 above)
Comment (galales) [reply privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/DynamoService.scala]: 👍

Comment (galales) [reply privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/DynamoService.scala]: or, if you want to avoid the use of the execution context
`value.map(_.some).leftMap(err => DynamoReadingError(describe(err))).toFuture`

File: project/Dependencies.scala
Comment (galales) [inline project/Dependencies.scala:1]: you should be able to remove also the `akka` object in this file

**Review by galales — APPROVED** [review]

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/Main.scala [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+import sttp.client4.httpclient.HttpClientFutureBackend
+import sttp.client4.{BackendOptions, WebSocketBackend}
+
+import it.pagopa.interop.privacynoticesupdater.util._
+import it.pagopa.interop.privacynoticesupdater.service._
+import it.pagopa.interop.privacynoticesupdater.converters.PrivacyNoticeConverter._
+
+import scala.concurrent.{Future, ExecutionContext}
+import scala.concurrent.Await
+import scala.concurrent.ExecutionContext.global
```
Comment (TonioGela) [inline privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/Main.scala]: If I'm not wrong every Future in this job gets scheduled on the global execution context, right?
That's quite an anti-pattern, so let's schedule a call to refactor it a bit together, using proper thread pools.

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/Main.scala [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    } yield (config, ots, ds, bk)
+  }
+
+  def execution(config: Configuration, ots: OneTrustService, ds: DynamoService)(implicit
+    global: ExecutionContext
+  ): Future[Unit] = for {
+    token  <- ots.getBearerToken()
+    ppOts  <- ots.getById(config.oneTrust.ppUuid)(token.access_token)
+    ppDs   <- ds.getById(config.oneTrust.ppUuid)
+    _      <- ppOts.fold(ppDs.fold(Future.successful(()))(p => ds.delete(p.pnId)))(p => ds.put(p.toPersistent))
```
Comment (TonioGela) [inline privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/Main.scala]: 🤯 2 nested folds are maybe too much :D

**Review by TonioGela — CHANGES_REQUESTED** [review]

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/DynamoService.scala [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      .exec(table.put(privacyNotice))
+  }
+
+  override def delete(id: UUID)(implicit contexts: Seq[(String, String)]): Future[Unit] = {
+    logger.info(s"Deleting privacy notice with id $id")
+    scanamo
+      .exec(
+        table
+          .delete("pk" === s"${PrivacyNotice.pkPrefix}$id" and "sk" === s"${PrivacyNotice.skPrefix}$id")
+      )
```
Comment (TonioGela) [inline privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/DynamoService.scala]: hadouken

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/OneTrustService.scala [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+          case Right(success) => Future.successful(Some(success))
+          case Left(error)    =>
+            error match {
+              case HttpError(body, StatusCode.NotFound) => {
+                logger.info(s"Receiving Not Found message")
+                if (body.map(_.path).isDefined)(Future.failed(OneTrustHttpError(StatusCode.NotFound)))
+                else (Future.successful(None))
+              }
+              case HttpError(_, statusCode)             => Future.failed(OneTrustHttpError(statusCode))
+              case DeserializationException(_, error)   => Future.failed(OneTrustDeserializationError(error))
```
Comment (TonioGela) [inline privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/OneTrustService.scala]: How about collapsing it in a single pattern match?

```suggestion
          case Left(HttpError(body, StatusCode.NotFound)) => {
                logger.info(s"Receiving Not Found message")
                if (body.map(_.path).isDefined)(Future.failed(OneTrustHttpError(StatusCode.NotFound)))
                else (Future.successful(None))
              }
          case Left(HttpError(_, statusCode))            => Future.failed(OneTrustHttpError(statusCode))
          case Left(DeserializationException(_, error))  => Future.failed(OneTrustDeserializationError(error))

```

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/OneTrustService.scala [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+      .send(backend)
+
+    sttpRequest.flatMap { response =>
+      response.body match {
+        case Right(success) => Future.successful(success)
+        case Left(error)    =>
+          error match {
+            case HttpError(body, statusCode)        => Future.failed(OneTrustAuthError(body, statusCode))
+            case DeserializationException(_, error) => Future.failed(OneTrustDeserializationError(error))
+          }
```
Comment (TonioGela) [inline privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/OneTrustService.scala]: ditto

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/Main.scala [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    ppOts  <- ots.getById(config.oneTrust.ppUuid)(token.access_token)
+    ppDs   <- ds.getById(config.oneTrust.ppUuid)
+    _      <- ppOts match {
+      case Some(p) => ds.put(p.toPersistent)
+      case None    =>
+        ppDs match {
+          case Some(p) => ds.delete(p.pnId)
+          case None    => Future.unit
+        }
+    }
```
Comment (TonioGela) [inline privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/Main.scala]: @galales @beetlecrunch 
I was trying to write this as follows

```scala
ppOts.map(p => ds.put(p.toPersistent))
    .orElse(ppDs.map(p => ds.delete(p.pnId)))
    .getOrElse(Future.unit)
```
but I'm not sure that's more readable. WDYT?

**Review by TonioGela — APPROVED** [review]

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/Main.scala (hunk 11 above)
Comment (galales) [reply privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/Main.scala]: the refactor works for me 👍 

I just find it a bit confusing and hard to follow due to the very compact and similar variables naming, which could be more transparent

File: privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/DynamoService.scala (hunk 8 above)
Comment (nttdata-rtorsoli) [reply privacy-notices-updater/src/main/scala/it/pagopa/interop/privacynoticesupdater/service/DynamoService.scala]: Have I to rewrite this without using Free Monad ScanamoOps?
